services/QuillLMS/client/app/bundles/Diagnostic/components/questions/response.tsx

```diff
@@ -0,0 +1,446 @@
+import * as React from 'react'
+import * as jsDiff from 'diff'
+
+import { ContentState, EditorState } from 'draft-js'
+import _ from 'underscore'
+
+import {
+  Modal,
+  TextEditor
+} from '../../../Shared/index'
+import massEdit from '../../actions/massEdit'
+import questionActions from '../../actions/questions'
+import {
+  deleteResponse,
+  getGradedResponsesWithCallback,
+  submitResponseEdit,
+} from '../../actions/responses'
+import sentenceFragmentActions from '../../actions/sentenceFragments'
+import C from '../../constants'
+import { rematchOne } from '../../libs/grading/rematching'
+import { getStatusForResponse } from '../../../Shared/index'
+
+import ConceptSelectorWithCheckbox from '../shared/conceptSelectorWithCheckbox.jsx'
+
+import ResponseList from './responseList'
+import { Question, Response } from '../../libs/grading/rematching'
+
+interface ResponseProps {
+  allExpanded: boolean,
+  ascending: boolean,
+  concepts: Array<Object>,
+  dispatch: Function,
+  expand: Function,
+  expanded: boolean,
+  getChildResponses: Function,
+  getResponse: Function,
+  mode: string,
+  massEditResponses: Object,
+  passedResponse: Response,
+  question: Question,
+  questionID: string,
+  readOnly: boolean,
+  responses: Array<Response>,
+  state: Object,
+  states: Object,
+  updateParentResponse: Function,
+}
+
+
+const Response = ({allExpanded, ascending, concepts, dispatch, expand, expanded, getChildResponses, getResponse, massEditResponses, mode, passedResponse, question, questionID, readOnly, responses, state, states, updateParentResponse}: ResponseProps) => {
+  const [response, setResponse] = React.useState(passedResponse)
+  const [feedback, setFeedback] = React.useState(passedResponse.feedback || '')
+  const [selectedBoilerplate, setSelectedBoilerplate] = React.useState('')
+  const [parent, setParent] = React.useState(null)
+
+  let conceptResultsTemp = {}
+  if (passedResponse.concept_results) {
+    if (typeof passedResponse.concept_results === 'string') {
+      conceptResultsTemp = JSON.parse(passedResponse.concept_results)
+    } else {
+      conceptResultsTemp = passedResponse.concept_results
+    }
+  }
+  const [conceptResults, setConceptResults] = React.useState(conceptResultsTemp)
```

Comment:
I would refactor this section so you have a function outside of the component that looks like this:

```
function extractConceptResultsFromResponse(response) {
  const { concept_results, } = response

  if (!concept_results) { return {} }

  if (typeof concept_results === 'string') {
    return JSON.parse(concept_results)
  } else {
    return concept_results
  }
}
```

And then line 64 can just be 

```
const [conceptResults, setConceptResults] = React.useState(extractConceptResultsFromResponse(passedResponse))
```